File: DSA/Linkedlist/Code/TraversalOfDoublyLinkedList.py
# Traversal of Doubly Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedListTraversal:
    def __init__(self):
        self.head = None
    
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
        new_node.prev = temp
    
    def display_forward(self):
        print("Forward: ", end="")
        temp = self.head
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print(" <-> ".join(elements) + " <-> NULL")
    
    def display_backward(self):
        if self.head is None:
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        print("Backward: ", end="")
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.prev
        print(" <-> ".join(elements) + " <-> NULL")
    
    def count_nodes(self):
        count = 0
        temp = self.head
        while temp is not None:
            count += 1
            temp = temp.next
        return count
    
    def get_node(self, pos):
        if pos < 1:
            return None
        temp = self.head
        for i in range(1, pos):
            if temp is None:
                return None
            temp = temp.next
        return temp

# Test
list1 = DoublyLinkedListTraversal()

print("=== Traversal of Doubly Linked List ===\n")

# Build list
for i in range(1, 6):
    list1.add_node(i * 10)

# Test 1: Forward traversal
list1.display_forward()

# Test 2: Backward traversal
list1.display_backward()

# Test 3: Count nodes
print(f"Total nodes: {list1.count_nodes()}")

# Test 4: Get at position
node = list1.get_node(3)
if node is not None:
    print(f"Node at position 3: {node.data}")

print("\nComplexity Analysis:")
print("Forward traversal: O(n)")
print("Backward traversal: O(n)")
print("Count nodes: O(n)")
print("Get at position: O(n)")

File: DSA/Linkedlist/Code/TraversalOfDoublyLinkedList.java
// Traversal of Doubly Linked List in Java

class Node {
    int data;
    Node next;
    Node prev;
    
    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class DoublyLinkedListTraversal {
    Node head;
    
    void addNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
        newNode.prev = temp;
    }
    
    void displayForward() {
        System.out.print("Forward: ");
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        }
        System.out.println("NULL");
    }
    
    void displayBackward() {
        if (head == null) return;
        
        Node temp = head;
        while (temp.next != null)
            temp = temp.next;
        
        System.out.print("Backward: ");
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.prev;
        }
        System.out.println("NULL");
    }
    
    int countNodes() {
        int count = 0;
        Node temp = head;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        return count;
    }
    
    Node getNode(int pos) {
        if (pos < 1) return null;
        Node temp = head;
        for (int i = 1; i < pos && temp != null; i++)
            temp = temp.next;
        return temp;
    }
    
    public static void main(String[] args) {
        DoublyLinkedListTraversal list = new DoublyLinkedListTraversal();
        
        System.out.println("=== Traversal of Doubly Linked List ===\n");
        
        // Build list
        for (int i = 1; i <= 5; i++)
            list.addNode(i * 10);
        
        // Test 1: Forward traversal
        list.displayForward();
        
        // Test 2: Backward traversal
        list.displayBackward();
        
        // Test 3: Count nodes
        System.out.println("Total nodes: " + list.countNodes());
        
        // Test 4: Get at position
        Node node = list.getNode(3);
        if (node != null)
            System.out.println("Node at position 3: " + node.data);
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Forward traversal: O(n)");
        System.out.println("Backward traversal: O(n)");
        System.out.println("Count nodes: O(n)");
        System.out.println("Get at position: O(n)");
    }
}
